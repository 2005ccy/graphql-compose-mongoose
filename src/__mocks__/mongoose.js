'use strict';

jest.disableAutomock();
module.exports = require.requireActual('mongoose');
jest.enableAutomock();
