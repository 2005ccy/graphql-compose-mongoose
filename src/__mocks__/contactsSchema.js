jest.unmock('mongoose');
import { Schema } from 'mongoose';

const ContactsSchema = new Schema(
  {
    phones: [String],
    email: String,
    skype: String,
  }
);

export default ContactsSchema;
