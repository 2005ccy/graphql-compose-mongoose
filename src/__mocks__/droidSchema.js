import { Schema } from 'mongoose';

export const DroidSchema = new Schema({
  makeDate: Date,
  modelNumber: Number,
  primaryFunction: [String],
});
