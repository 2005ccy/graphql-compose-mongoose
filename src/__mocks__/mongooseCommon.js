import mongoose, { Schema } from 'mongoose';

const dbName = 'gqc-mongoose-test';
const uri = `mongodb://127.0.0.1:27017/${dbName}`;

mongoose.connect(uri);

mongoose.connection.on('error', (err) => {
  throw new Error(err);
});


function dropDBs(done) {
  mongoose.connection.db.dropDatabase(() => {
    if (done) {
      done();
    }
  });
}

export {
  mongoose,
  Schema,
  dropDBs,
  uri,
  dbName,
};
