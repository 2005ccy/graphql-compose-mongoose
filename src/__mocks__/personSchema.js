/* @flow */

import { Schema } from './mongooseCommon';

export const PersonSchema = new Schema({
  dob: Number,
  starShips: [String],
  totalCredits: Number,
});
