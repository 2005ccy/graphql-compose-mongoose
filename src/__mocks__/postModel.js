/* @flow */

import type { Schema as SchemaType } from 'mongoose';
import { mongoose, Schema } from './mongooseCommon';

const PostSchema: SchemaType<any> = new Schema({
  _id: {
    type: Number,
  },
  title: {
    type: String,
    description: 'Post title',
  },

  // createdAt, created via option `timastamp: true` (see bottom)
  // updatedAt, created via option `timastamp: true` (see bottom)
});

const PostModel = mongoose.model('Post', PostSchema);

export { PostSchema, PostModel };
