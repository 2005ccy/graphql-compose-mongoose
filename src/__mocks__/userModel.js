import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguagesSchema from './languagesSchema';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'UserModel',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'UserModel',
    },

    name: {
      type: String,
      description: 'Person name',
    },

    age: {
      type: Number,
      description: 'Full years',
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [{
        type: String,
        enum: Object.keys(enumEmployment),
      }],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguagesSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    // createdAt, created via option `timastamp: true`

    // updatedAt, created via option `timastamp: true`

    __secretField: {
      type: String,
    },
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

UserSchema.virtual('nameVirtual').get(function () { // eslint-disable-line
  return `VirtualFieldValue${this._id}`;
});


const UserModel = mongoose.model('UserModel', UserSchema);

export {
  UserSchema,
  UserModel,
};
