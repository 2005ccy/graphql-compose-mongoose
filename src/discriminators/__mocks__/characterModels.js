/* @flow */

import type { MongooseModel } from 'mongoose';
import { mongoose, Schema, Types } from '../../__mocks__/mongooseCommon';
import { DroidSchema } from './droidSchema';
import { PersonSchema } from './personSchema';

const enumCharacterType = {
  PERSON: 'Person',
  DROID: 'Droid',
};

export const CharacterObject = {
  _id: {
    type: String,
    default: () => new Types.ObjectId(),
  },
  name: String,

  type: {
    type: String,
    require: true,
    enum: (Object.keys(enumCharacterType): Array<string>),
  },
  kind: {
    type: String,
    require: true,
    enum: (Object.keys(enumCharacterType): Array<string>),
  },

  friends: [String], // another Character
  appearsIn: [String], // movie
};

const CharacterSchema = new Schema(CharacterObject);
const ACharacterSchema = new Schema(Object.assign({}, CharacterObject));

export function getCharacterModels(DKey: string) {
  CharacterSchema.set('discriminatorKey', DKey);

  const CharacterModel: MongooseModel = mongoose.models.Character
    ? mongoose.models.Character
    : mongoose.model('Character', CharacterSchema);

  const PersonModel: MongooseModel = mongoose.models[enumCharacterType.PERSON]
    ? mongoose.models[enumCharacterType.PERSON]
    : CharacterModel.discriminator(enumCharacterType.PERSON, PersonSchema);

  const DroidModel: MongooseModel = mongoose.models[enumCharacterType.DROID]
    ? mongoose.models[enumCharacterType.DROID]
    : CharacterModel.discriminator(enumCharacterType.DROID, DroidSchema);

  return { CharacterModel, PersonModel, DroidModel };
}

export function getCharacterModelClone() {
  const NoDKeyCharacterModel = mongoose.model('NoDKeyCharacter', ACharacterSchema);

  /*
    const APersonModel = ACharacterModel.discriminator('A' + enumCharacterType.PERSON, PersonSchema.clone());

    const ADroidModel = ACharacterModel.discriminator('A' + enumCharacterType.DROID, DroidSchema.clone());
  */

  return { NoDKeyCharacterModel }; // APersonModel, ADroidModel };
}
