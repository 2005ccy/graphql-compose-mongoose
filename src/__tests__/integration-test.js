/* @flow */

import { GQC } from 'graphql-compose';
import { graphql } from 'graphql-compose/lib/graphql';
import { UserModel } from '../__mocks__/userModel';
import { composeWithMongoose } from '../index';
import typeStorage from '../typeStorage';

beforeAll(() => UserModel.base.connect());
afterAll(() => UserModel.base.disconnect());

describe('integration tests', () => {
  beforeEach(() => {
    typeStorage.clear();
    UserModel.schema._gqcTypeComposer = undefined;
  });

  describe('projection', async () => {
    let schema;
    let UserTC;
    beforeAll(async () => {
      UserTC = composeWithMongoose(UserModel);
      UserTC.addFields({
        rawData: {
          type: 'JSON',
          resolve: source => source.toJSON(),
          projection: { '*': true },
        },
      });
      GQC.rootQuery().addFields({
        user: UserTC.getResolver('findById'),
      });
      schema = GQC.buildSchema();
      await UserModel.create({
        _id: '100000000000000000000000',
        name: 'Name',
        age: 20,
        gender: 'male',
        skills: ['a', 'b', 'c'],
        relocation: true,
      });
    });

    it('should request only fields from query', async () => {
      const res = await graphql(schema, '{ user(_id: "100000000000000000000000") { name } }');
      expect(res).toMatchSnapshot('projection from query fields');
    });

    it('should request all fields to rawData field', async () => {
      const res: any = await graphql(
        schema,
        '{ user(_id: "100000000000000000000000") { rawData } }'
      );
      expect(Object.keys(res.data.user.rawData)).toMatchSnapshot('projection from all fields');
    });
  });
});
