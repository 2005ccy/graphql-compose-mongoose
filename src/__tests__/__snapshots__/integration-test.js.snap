// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projection from all fields 1`] = `
Array [
  "_id",
  "updatedAt",
  "createdAt",
  "name",
  "age",
  "gender",
  "relocation",
  "__v",
  "languages",
  "contacts",
  "employment",
  "skills",
  "users",
  "nameVirtual",
  "id",
]
`;

exports[`projection from query fields 1`] = `
Object {
  "data": Object {
    "user": Object {
      "name": "Name",
    },
  },
}
`;
